File: 0859-design-circular-deque/0859-design-circular-deque.cs
public class Queue
{
    
    public Queue next;
    public Queue prev;
    public int val;
    public Queue(int v)
    {
        val = v;
        next = null;
        prev = null;
    }
}

public class MyCircularDeque {

    public int size;
    public int capacity;
    Queue front;
    Queue rear;

    public MyCircularDeque(int k) {
        size = 0;
        capacity = k;
        front = null;
        rear = null;
    }
    
    public bool InsertFront(int value) {
        if(IsFull())
        {
            return false;
        }
        if(IsEmpty())
        {
            Queue newFront = new Queue(value);
            front = newFront;
            rear = front;
        }
        else
        {
            Queue newFront = new Queue(value);
            newFront.next = front;
            front.prev = newFront;
            front = newFront;
        }
        size += 1;
        return true;
    }
    
    public bool InsertLast(int value) {
        if(IsFull())
        {
            return false;
        }
        if(IsEmpty())
        {
            Queue newFront = new Queue(value);
            front = newFront;
            rear = front;
        }
        else
        {
            Queue newRear = new Queue(value);
            newRear.prev = rear;
            rear.next = newRear;
            rear = newRear;
        }
        size += 1;
        return true;
    }
    
    public bool DeleteFront() {
        if(IsEmpty())
        {
            return false;
        }

        Queue temp = front;
        front = front.next;

        temp.next = null;
        temp.prev = null;
        temp = null;

        size -= 1;

        if(IsEmpty())
        {
            front = null;
            rear = null;
        }
        return true;
    }
    
    public bool DeleteLast() {
        if(IsEmpty())
        {
            return false;
        }

        Queue temp = rear;
        rear = rear.prev;

        temp.next = null;
        temp.prev = null;
        temp = null;

        size -= 1;

        if(IsEmpty())
        {
            front = null;
            rear = null;
        }
        return true;
    }
    
    public int GetFront() {
        if(IsEmpty())
        return -1;
        return front.val;
    }
    
    public int GetRear() {
        if(IsEmpty())
        {
            return -1;
        }
        return rear.val;
    }
    
    public bool IsEmpty() {
        if(size == 0)
        return true;
        return false;
    }
    
    public bool IsFull() {
        if(size == capacity)
        return true;
        return false;
    }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque obj = new MyCircularDeque(k);
 * bool param_1 = obj.InsertFront(value);
 * bool param_2 = obj.InsertLast(value);
 * bool param_3 = obj.DeleteFront();
 * bool param_4 = obj.DeleteLast();
 * int param_5 = obj.GetFront();
 * int param_6 = obj.GetRear();
 * bool param_7 = obj.IsEmpty();
 * bool param_8 = obj.IsFull();
 */

File: 0859-design-circular-deque/0859-design-circular-deque.cpp
class DQueue
{
    public:
    DQueue * next;
    DQueue * rear;
    int val;

    DQueue(int v)
    {
        val = v;
        next = NULL;
        rear = NULL;
    }
};

class MyCircularDeque {
public:
    DQueue * headFront;
    DQueue * headRear;
    int size;
    int maxSize;

    MyCircularDeque(int k) {
        headFront = NULL;
        headRear = NULL;
        size = 0;
        maxSize = k;
    }
    
    bool insertFront(int value) {
        if(size == maxSize)
            return false;

        DQueue * temp = new DQueue(value);
        if(headFront == NULL)
        {
            headRear = temp;
            headFront = temp;
        }
        else
        {
            headFront->rear = temp;
            temp->next = headFront;
            headFront = temp;
        }

        
        size += 1;
        return true;
    }
    
    bool insertLast(int value) {
        if(size == maxSize)
            return false;

        DQueue * temp = new DQueue(value);
        if(headRear == NULL)
        {
            headRear = temp;
            headFront = temp;
        }
        else
        {
            headRear->next = temp;
            temp->rear = headRear;
            headRear = temp;
        }
        size += 1;
        return true;
    }
    
    bool deleteFront() {
        if(headFront == NULL)
            return false;

        DQueue * temp = headFront;
        headFront = temp->next;
        temp = NULL;
        delete temp;
        size -= 1;

        if(size == 0)
        {
            headFront = NULL;
            headRear = NULL;
        }
        return true;
    }
    
    bool deleteLast() {
        if(headRear == NULL)
            return false;

        DQueue * temp = headRear;
        headRear = temp->rear;
        temp = NULL;
        delete temp;

        size -= 1;

         if(size == 0)
        {
            headFront = NULL;
            headRear = NULL;
        }
        return true;
    }
    
    int getFront() {
        return headFront == NULL ? -1 :  headFront->val;
    }
    
    int getRear() {
        return headRear == NULL ? -1 : headRear->val;
    }
    
    bool isEmpty() {
        return size == 0 ? true:false;
    }
    
    bool isFull() {
        return size == maxSize ? true:false;
    }
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque* obj = new MyCircularDeque(k);
 * bool param_1 = obj->insertFront(value);
 * bool param_2 = obj->insertLast(value);
 * bool param_3 = obj->deleteFront();
 * bool param_4 = obj->deleteLast();
 * int param_5 = obj->getFront();
 * int param_6 = obj->getRear();
 * bool param_7 = obj->isEmpty();
 * bool param_8 = obj->isFull();
 */